### PR TITLE
Debugging Feature for Alexa SDK

### PR DESCRIPTION
**Pull request by allusai**
## Description

This is a feature that I wrote for the Alexa SDK that provides a simple, clean interface to help you debug your code. This library makes debugging as simple as writing one line of code. Using these functions, you can actually hear your debug statements through Alexa which makes interacting with debugging code much more elegant. You retain the option of seeing your debugging text in the terminal as well if you would like. So there's nothing to lose when using this feature, and you no longer have to write as much uninteresting debugging code.

## Motivation and Context

Currently debugging requires you to switch back and forth between the developer console 
where you give commands, AWS Lambda, and the terminal where you see your print statements. 
Also, writing code to help read out values for objects or slots is not a trivial task and 
takes away time from actually working on the skill itself.

## Testing

I have tested this feature on two of my own Alexa Skill projects and everything works smoothly. 
I have made sure to add helpful error handling as well for mistakes the user of this feature 
might make. Since this feature is doesn't affect or interact with any other files in the SDK, it is 
secure to use.

[Getting-Started-With-Alexa-Speech-Debugging.md.zip](https://github.com/alexa/alexa-skills-kit-sdk-for-nodejs/files/2112942/Getting-Started-With-Alexa-Speech-Debugging.md.zip)

## Screenshots (if appropriate)
Code for test:
```
    handlerInput.requestEnvelope.request.intent.slots['PetSlot'].value = 'Dragon';
    debug.start();
    debug.speakSlots(handlerInput);
    return debug.complete(handlerInput);
```
Results:
slots test

Code for test:
```
    debug.start();
    debug.speak("Value of x is: " + 17);
    debug.speak(true);
    
    var car = {brand:"Fiat", model:"500", color:"white"};
    debug.speak(car);
    
    return debug.complete(handlerInput);
```
Results:
magpie test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/alexa/alexa-skills-kit-sdk-for-nodejs/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 node_modules/
 package-lock.json
 lerna-debug.log
+.DS_Store
```
Comment (tianrenz) [inline .gitignore]: This is OS specific item that should be in the global gitignore.

Comment (allusai) [reply .gitignore]: Alright, I will make that change.

**Comment by allusai** [comment]
Of course! I will try and do that as soon as possible. Thank you for the consideration.

**Comment by allusai** [comment]
Hello @tianrenz ,

I am in the process of rewriting my feature into Typescript and I just had a few design questions that I wanted to get your opinion on. Would you say that it is better style to write a Typescript file that exports each individual function and then in the runner file have an "import * as debug from './Debug' " (like the attached file) or instead write an interface/class for the feature? I've noticed that some of the other code files (like ResponseBuilder.ts) generally do one interface Typescript file and one implementation file, but I think my feature might work slightly differently from those.

And I know the ASK SDK is implemented in TypeScript, but I noticed that a lot of the Alexa Skill templates on AWS Lambda instead use a JavaScript file in conjunction with a ".d.ts" declaration file. Would you like me to also write this second set of files so that the feature is available to users who have that setup?

Thank you again, I just want make sure this feature integrates cleanly with the project!

**Comment by tianrenz** [comment]
Hi, 

My recommendation is to wrap all of the functions into a module, to make it modular and extensible. Also, have you considered maybe releasing this as an separate NPM package? This can be a pretty good utility package that people can import and use on their own. 

For the file types, when you compile the typescript file, make sure to have ` "declaration" : true,` so the compiler will generate the .d.ts file for you

Regards

**Comment by allusai** [comment]
Good afternoon @tianrenz,

I have followed your suggestions and converted the file to Typescript. I have also run all the necessary tests to make sure everything still works.

I looked into the idea of making this a separate npm package and my reason for thinking otherwise is that debugging is really central to developing an Alexa skill. I believe if this feature was included as part of the SDK, then more people would have access to it. Whereas if this were a separate package, I think people would have a hard time finding it or even knowing about its existence. What do you think?

Best Wishes,
Sai

**Comment by allusai** [comment]
Good evening @tianrenz,

I have documented this feature thoroughly in the attached file. If you wish, I can also provide you with the test suite that I used to verify that the program works in nominal and edge cases. If you would like anything else from my end, please let me know!

Thank you,
Sai

[Getting-Started-With-Alexa-Speech-Debugging.md.zip](https://github.com/alexa/alexa-skills-kit-sdk-for-nodejs/files/2129540/Getting-Started-With-Alexa-Speech-Debugging.md.zip)

**Comment by allusai** [comment]
Good afternoon @tianrenz,

Thank you so much! I'm excited at the prospect that my work could assist thousands of Alexa developers like myself with an issue that they commonly face.

Best Wishes,
Sai

**Comment by allusai** [comment]
Hello @tianrenz,

I hope you are doing well. Is there anything else you need from me for the review?

Thank you,
Sai

**Comment by tianrenz** [comment]
Hi @allusai , 

I think the changes you purposed here would make a great utility library for developers to adopt. But it doesn't really fit our requirement for SDK debugging support feature as it needs to alter the skill behavior to read out user input, which is something that's not encouraged by Amazon data security policy. We would still recommend to release this on your own for a utility package in order to share with the community. 

Regards
